File: src/swift/graph/dfs/IterativeTraversal.swift
typealias Graph<T: Hashable> = [T: [T]]


func traverse<T>(in graph: Graph<T>, withRoot root: T) -> [T]? {
    guard graph[root] != nil else { return nil }

    var explored = Set<T>()
    var result = [T]()
    var stack = [root]

    while !stack.isEmpty {
        let vertex = stack.removeLast()

        if !explored.contains(vertex) {
            explored.insert(vertex)
            result.append(vertex)

            if let successors = graph[vertex] {
                stack += successors
            }
        }
    }
    return result
}


func main() {
    let graph = [
        // first subgraph
        "A": ["B", "C", "D"],
        "B": ["E"],
        "C": ["F", "G"],
        "F": ["H", "I"],

        // second subgraph
        "J": ["K", "L", "M"],
        "L": ["N", "O"],       
    ]

    if let result = traverse(in: graph, withRoot: "A") {
        print(result) // ["A", "D", "C", "G", "F", "I", "H", "B", "E"]
    }
}

File: src/java/graph/dfs/IterativeTraversal.java
package graph.dfs;

import java.util.ArrayDeque;
import java.util.Collection;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;


public class IterativeTraversal {

    public static <T> List<T> traverse(Map<T, List<T>> graph, T root) {
        if (!graph.containsKey(root))
            throw new IllegalArgumentException("Invalid root vertex");

        var explored = new LinkedHashSet<T>();
        var stack = new ArrayDeque<T>();
        stack.add(root);

        while (!stack.isEmpty()) {
            var vertex = stack.removeLast();
            if (!explored.contains(vertex)) {
                explored.add(vertex);
                stack.addAll(graph.getOrDefault(vertex, List.of()));
            }
        }
        
        return List.copyOf(explored);
    }


    public static void main(String[] args) {
        var graph = Map.of(
                // first subgraph
                "A", List.of("B", "C", "D"),
                "B", List.of("E"),
                "C", List.of("F", "G"),
                "F", List.of("H", "I"),

                // second subgraph
                "J", List.of("K", "L", "M"),
                "L", List.of("N", "O")
        );

        var result = traverse(graph, "A");

        System.out.println(result); // [A, D, C, G, F, I, H, B, E]
    }
}